python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -2394,6 +2399,42 @@ def get_handlers(self) -> Sequence[logging.Handler]:
     def store_event(self, event: "EventLogEntry") -> None:
         self._event_storage.store_event(event)
 
+    def should_retry_run(self, run: DagsterRun) -> bool:
```

Comment:
I think it's worth adding a hefty docstring here explaining in english the cases where we will and will not retry. It gets pretty complex and would be nice to not need to rederive that in the future.